tensorflow/core/kernels/sparse_to_dense_op_gpu.cu.cc

```diff
@@ -0,0 +1,199 @@
+// check passes. Otherwise, it represents the first detected position of the
+// invalid index for the check.
+struct IndicesValidStatus {
+  int valid;
+  int increasing;
+  int different;
+};
+
+}  // namespace
+
+namespace functor {
+
+template <typename T, typename Index>
+void LaunchSparseToDense<T, Index>::operator()(
+    OpKernelContext* c, AsyncOpKernel::DoneCallback done, AsyncOpKernel* op,
+    bool validate_indices, const Index* indices, const T* values,
+    const int num_elems, const int num_values, const Index* shape,
+    const int num_dims, const T default_value, Index dense_size, T* dense) {
+  auto* stream = c->op_device_context()->stream();
+  const Eigen::GpuDevice& d = c->eigen_gpu_device();
+
+  if (validate_indices && num_elems != 0) {
+    VLOG(1) << "SparseToDense will be performed on GPUs. For performance "
+        "reasons, it is suggested to pass False to validate_indices.";
+
+    IndicesValidStatus valid_status;
+    int valid_status_size = sizeof(valid_status) / sizeof(int);
+    int valid_status_bytes = sizeof(valid_status);
+
+    Tensor valid_status_tensor;
+    OP_REQUIRES_OK_ASYNC(c, c->allocate_temp(
+        DT_INT32, TensorShape({valid_status_size}),
+        &valid_status_tensor), done);
+
+    auto status_ptr = valid_status_tensor.template flat<int>().data();
+    se::DeviceMemoryBase valid_status_ptr(status_ptr, valid_status_bytes);
+
+    GpuLaunchConfig config = GetGpuLaunchConfig(num_elems, d);
+    stream->ThenMemset32(&valid_status_ptr, INT_MAX, valid_status_bytes);
+    OP_REQUIRES_OK_ASYNC(c, GpuLaunchKernel(
+        CheckIndicesValid<Index>, config.block_count, config.thread_per_block,
+        0, d.stream(), indices, num_elems, shape, num_dims, status_ptr), done);
+    stream->ThenMemcpy(reinterpret_cast<int*>(&valid_status), valid_status_ptr,
+                       valid_status_bytes);
+
+    auto check_status = [op, c, &valid_status, done]() {
+      OP_REQUIRES_ASYNC(c, valid_status.valid == INT_MAX,
+                        errors::InvalidArgument(
+                            "indices[", valid_status.valid,
+                            "] is out of bounds."), done);
+
+      OP_REQUIRES_ASYNC(c, valid_status.increasing == INT_MAX,
+                        errors::InvalidArgument(
+                            "indices[", valid_status.increasing, "] is out of "
+                            "order. Many sparse ops require sorted indices.\n"
+                            "  Use `tf.sparse.reorder` to create a correctly "
+                            "ordered copy.\n\n"), done);
+
+      OP_REQUIRES_ASYNC(c, valid_status.different == INT_MAX,
+                        errors::InvalidArgument(
+                            "indices[", valid_status.different, "] is "
+                            "repeated."),
+                        done);
+      done();
+    };
+
+    c->device()->tensorflow_gpu_device_info()->event_mgr->ThenExecute(
+        stream, check_status);
```

Comment:
Maybe I'm missing something, but won't we end up calling `done` twice if we take this `validate_indices && num_elems != 0` branch?  Once after the d2h is finished and once in line 186?